test/test_datasets.py

```diff
@@ -280,6 +283,92 @@ def test_ucf101(self):
+
+        def target_transform(target):
+            return expected_target
+
+        with places365_root() as places365:
+            root, data = places365
+
+            dataset = torchvision.datasets.Places365(
+                root, transform=transform, target_transform=target_transform, download=True
+            )
+            actual_image, actual_target = dataset[0]
+
+            self.assertEqual(actual_image, expected_image)
+            self.assertEqual(actual_target, expected_target)
+
+    @mock.patch("torchvision.datasets.utils.download_url")
+    def test_places365_downloadable(self, download_url):
+        for split, small in itertools.product(("train-standard", "train-challenge", "val", "test"), (False, True)):
+            with places365_root(split=split, small=small) as places365:
+                root, data = places365
+
+                torchvision.datasets.Places365(root, split=split, small=small, download=True)
+
+        urls = {call_args[0][0] for call_args in download_url.call_args_list}
+        for url in urls:
+            with self.subTest(url=url):
+                response = urlopen(Request(url, method="HEAD"))
+                assert response.code == 200, f"Server returned status code {response.code} for {url}."
+
+    def test_places365_devkit_download(self):
+        for split in ("train-standard", "train-challenge", "val", "test"):
+            with self.subTest(split=split):
+                with places365_root(split=split) as places365:
+                    root, data = places365
+
+                    dataset = torchvision.datasets.Places365(root, split=split, download=True)
+
+                    with self.subTest("classes"):
+                        self.assertSequenceEqual(dataset.classes, data["classes"])
+
+                    with self.subTest("class_to_idx"):
+                        self.assertDictEqual(dataset.class_to_idx, data["class_to_idx"])
+
+                    with self.subTest("imgs"):
+                        self.assertSequenceEqual(dataset.imgs, data["imgs"])
+
+    def test_places365_devkit_no_download(self):
+        for split in ("train-standard", "train-challenge", "val", "test"):
+            with self.subTest(split=split):
+                with places365_root(split=split, extract_images=False) as places365:
+                    root, data = places365
+
+                    with self.assertRaises(RuntimeError):
+                        torchvision.datasets.Places365(root, split=split, download=False)
+
+    def test_places365_images_download(self):
+        for split, small in itertools.product(("train-standard", "train-challenge", "val", "test"), (False, True)):
+            with self.subTest(split=split, small=small):
+                with places365_root(split=split, small=small) as places365:
+                    root, data = places365
+
+                    dataset = torchvision.datasets.Places365(root, split=split, small=small, download=True)
```

Comment:
Just to double-check, we don't download any files here, right?

Reply:
Correct. We need to set `download=True` to enable the fake downloading. Since we are patching the class attributes, i.e. the MD5 checksums of the archives, `download_and_extract_url` is basically `extract_archive`.